File: packages/wasmvm/wasmlib/src/sandbox.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::rc::Rc;

use crate::*;
use crate::host::*;
use crate::wasmrequests::*;

// @formatter:off
pub const FN_ACCOUNT_ID               : i32 = -1;
pub const FN_ALLOWANCE                : i32 = -2;
pub const FN_BALANCE                  : i32 = -3;
pub const FN_BALANCES                 : i32 = -4;
pub const FN_BLOCK_CONTEXT            : i32 = -5;
pub const FN_CALL                     : i32 = -6;
pub const FN_CALLER                   : i32 = -7;
pub const FN_CHAIN_ID                 : i32 = -8;
pub const FN_CHAIN_OWNER_ID           : i32 = -9;
pub const FN_CONTRACT                 : i32 = -10;
pub const FN_DEPLOY_CONTRACT          : i32 = -11;
pub const FN_ENTROPY                  : i32 = -12;
pub const FN_ESTIMATE_STORAGE_DEPOSIT : i32 = -13;
pub const FN_EVENT                    : i32 = -14;
pub const FN_LOG                      : i32 = -15;
pub const FN_MINTED                   : i32 = -16;
pub const FN_PANIC                    : i32 = -17;
pub const FN_PARAMS                   : i32 = -18;
pub const FN_POST                     : i32 = -19;
pub const FN_REQUEST                  : i32 = -20;
pub const FN_REQUEST_ID               : i32 = -21;
pub const FN_REQUEST_SENDER           : i32 = -22;
pub const FN_RESULTS                  : i32 = -23;
pub const FN_SEND                     : i32 = -24;
pub const FN_STATE_ANCHOR             : i32 = -25;
pub const FN_TIMESTAMP                : i32 = -26;
pub const FN_TRACE                    : i32 = -27;
pub const FN_TRANSFER_ALLOWED         : i32 = -28;
pub const FN_UTILS_BECH32_DECODE      : i32 = -29;
pub const FN_UTILS_BECH32_ENCODE      : i32 = -30;
pub const FN_UTILS_BLS_ADDRESS        : i32 = -31;
pub const FN_UTILS_BLS_AGGREGATE      : i32 = -32;
pub const FN_UTILS_BLS_VALID          : i32 = -33;
pub const FN_UTILS_ED25519_ADDRESS    : i32 = -34;
pub const FN_UTILS_ED25519_VALID      : i32 = -35;
pub const FN_UTILS_HASH_BLAKE2B       : i32 = -36;
pub const FN_UTILS_HASH_NAME          : i32 = -37;
pub const FN_UTILS_HASH_SHA3          : i32 = -38;
// @formatter:on

// Direct logging of informational text to host log
pub fn log(text: &str) {
    sandbox(FN_LOG, text.as_bytes());
}

// Direct logging of error to host log, followed by panicking out of the Wasm code
pub fn panic(text: &str) {
    sandbox(FN_PANIC, text.as_bytes());
}

// Direct logging of debug trace text to host log
pub fn trace(text: &str) {
    sandbox(FN_TRACE, text.as_bytes());
}

pub fn params_proxy() -> Proxy {
    let buf = sandbox(FN_PARAMS, &[]);
    Proxy::new(&Rc::new(ScDict::new(&buf)))
}

pub fn results_proxy() -> Proxy {
    Proxy::new(&Rc::new(ScDict::new(&[])))
}

pub fn state_proxy() -> Proxy {
    Proxy::new(&Rc::new(ScDict::state()))
}

pub trait ScSandbox {
    // retrieve the agent id of this contract account
    fn account_id(&self) -> ScAgentID {
        agent_id_from_bytes(&sandbox(FN_ACCOUNT_ID, &[]))
    }

    fn balance(&self, token_id: ScTokenID) -> u64 {
        uint64_from_bytes(&sandbox(FN_BALANCE, &token_id.to_bytes()))
    }

    // access the current balances for all assets
    fn balances(&self) -> ScBalances {
        ScAssets::new(&sandbox(FN_BALANCES, &[])).balances()
    }

    // calls a smart contract function
    fn call_with_allowance(&self, h_contract: ScHname, h_function: ScHname, params: Option<ScDict>, allowance: Option<ScTransfer>) -> ScImmutableDict {
        let mut req = wasmrequests::CallRequest {
            contract: h_contract,
            function: h_function,
            params: vec![0; SC_UINT32_LENGTH],
            allowance: vec![0; SC_UINT32_LENGTH],
        };
        if let Some(params) = params {
            req.params = params.to_bytes();
        }
        if let Some(allowance) = allowance {
            req.allowance = allowance.to_bytes();
        }
        let buf = self.fn_call(&req);
        ScImmutableDict::new(ScDict::new(&buf))
    }

    fn fn_call(&self, req: &CallRequest) -> Vec<u8> {
        sandbox(FN_CALL, &req.to_bytes())
    }

    // retrieve the agent id of the owner of the chain this contract lives on
    fn chain_owner_id(&self) -> ScAgentID {
        agent_id_from_bytes(&sandbox(FN_CHAIN_OWNER_ID, &[]))
    }

    // retrieve the hname of this contract
    fn contract(&self) -> ScHname {
        hname_from_bytes(&sandbox(FN_CONTRACT, &[]))
    }

    // retrieve the chain id of the chain this contract lives on
    fn current_chain_id(&self) -> ScChainID {
        self.fn_chain_id()
    }

    fn fn_chain_id(&self) -> ScChainID {
        chain_id_from_bytes(&sandbox(FN_CHAIN_ID, &[]))
    }

    // logs informational text message
    fn log(&self, text: &str) {
        log(text);
    }

    // logs error text message and then panics
    fn panic(&self, text: &str) {
        panic(text);
    }

    // retrieve parameters passed to the smart contract function that was called
    fn params(&self) -> ScImmutableDict {
        let buf = sandbox(FN_PARAMS, &[]);
        ScImmutableDict::new(ScDict::new(&buf))
    }

    // panics if condition is not satisfied
    fn require(&self, cond: bool, msg: &str) {
        if !cond {
            panic(msg)
        }
    }

    fn results(&self, results: &Proxy) {
        sandbox(FN_RESULTS, &results.kv_store.to_bytes());
    }

    // deterministic time stamp fixed at the moment of calling the smart contract
    fn timestamp(&self) -> u64 {
        uint64_from_bytes(&sandbox(FN_TIMESTAMP, &[]))
    }

    // logs debugging trace text message
    fn trace(&self, text: &str) {
        trace(text);
    }

    // access diverse utility functions
    fn utility(&self) -> ScSandboxUtils {
        ScSandboxUtils {}
    }
}

pub trait ScSandboxView: ScSandbox {
    // calls a smart contract view
    fn call(&self, h_contract: ScHname, h_function: ScHname, params: Option<ScDict>) -> ScImmutableDict {
        return self.call_with_allowance(h_contract, h_function, params, None);
    }

    fn raw_state(&self) -> ScImmutableDict {
        ScImmutableDict::new(ScDict::state())
    }
}

pub trait ScSandboxFunc: ScSandbox {
    // access the allowance assets
    fn allowance(&self) -> ScBalances {
        let buf = sandbox(FN_ALLOWANCE, &[]);
        return ScAssets::new(&buf).balances();
    }

    //fn blockContext(&self, construct func(sandbox: ScSandbox) interface{}, onClose func(interface{})) -> interface{} {
    //	panic("implement me")
    //}

    // calls a smart contract func or view
    fn call(&self, h_contract: ScHname, h_function: ScHname, params: Option<ScDict>, allowance: Option<ScTransfer>) -> ScImmutableDict {
        return self.call_with_allowance(h_contract, h_function, params, allowance);
    }

    // retrieve the agent id of the caller of the smart contract
    fn caller(&self) -> ScAgentID {
        return agent_id_from_bytes(&sandbox(FN_CALLER, &[]));
    }

    // deploys a smart contract
    fn deploy_contract(&self, program_hash: &ScHash, name: &str, description: &str, init_params: Option<ScDict>) {
        let mut req = wasmrequests::DeployRequest {
            prog_hash: program_hash.clone(),
            name: name.to_string(),
            description: description.to_string(),
            params: vec![0; SC_UINT32_LENGTH],
        };
        if let Some(init_params) = init_params {
            req.params = init_params.to_bytes();
        }
        sandbox(FN_DEPLOY_CONTRACT, &req.to_bytes());
    }

    // returns random entropy data for current request.
    fn entropy(&self) -> ScHash {
        return hash_from_bytes(&sandbox(FN_ENTROPY, &[]));
    }

    fn estimate_storage_deposit(&self, f: &ScFunc) -> u64 {
        let req = f.post_request(ScFuncContext {}.current_chain_id());
        uint64_from_bytes(&sandbox(FN_ESTIMATE_STORAGE_DEPOSIT, &req.to_bytes()))
    }

    // signals an event on the node that external entities can subscribe to
    fn event(&self, msg: &str) {
        sandbox(FN_EVENT, &string_to_bytes(msg));
    }

    // retrieve the assets that were minted in this transaction
    fn minted(&self) -> ScBalances {
        let buf = sandbox(FN_MINTED, &[]);
        return ScAssets::new(&buf).balances();
    }

    // Post (delayed) posts a SC function request
    fn post(&self, chain_id: ScChainID, h_contract: ScHname, h_function: ScHname, params: ScDict, allowance: ScTransfer, transfer: ScTransfer, delay: u32) {
        let req = wasmrequests::PostRequest {
            chain_id,
            contract: h_contract,
            function: h_function,
            params: params.to_bytes(),
            allowance: allowance.to_bytes(),
            transfer: transfer.to_bytes(),
            delay: delay,
        };
        self.fn_post(&req);
    }

    fn fn_post(&self, req: &PostRequest) -> Vec<u8> {
        sandbox(FN_POST, &req.to_bytes())
    }

    // generates a random value from 0 to max (exclusive: max) using a deterministic RNG
    fn random(&self, max: u64) -> u64 {
        if max == 0 {
            panic("random: max parameter should be > 0");
        }
        unsafe {
            static mut ENTROPY: Vec<u8> = Vec::new();
            static mut OFFSET: usize = 0;
            // note that entropy gets reset for every request
            if ENTROPY.len() == 0 {
                // first time in self: request, initialize with current request entropy
                ENTROPY = self.entropy().to_bytes();
                OFFSET = 0;
            }
            if OFFSET == 32 {
                // ran out of entropy: data, hash entropy for next pseudo-random entropy
                ENTROPY = self.utility().hash_blake2b(&ENTROPY).to_bytes();
                OFFSET = 0;
            }
            let rnd = uint64_from_bytes(&ENTROPY[OFFSET..OFFSET + 8]) % max;
            OFFSET += 8;
            return rnd;
        }
    }

    fn raw_state(&self) -> ScDict {
        ScDict::state()
    }

    //fn request(&self) -> ScRequest {
    //	panic("implement me");
    //}

    // retrieve the request id of this transaction
    fn request_id(&self) -> ScRequestID {
        return request_id_from_bytes(&sandbox(FN_REQUEST_ID, &[]));
    }

    // retrieve the request sender of this transaction
    fn request_sender(&self) -> ScAgentID {
        return agent_id_from_bytes(&sandbox(FN_REQUEST_SENDER, &[]));
    }

    // Send transfers SC assets to the specified address
    fn send(&self, address: &ScAddress, transfer: &ScTransfer) {
        // we need some assets to send
        if transfer.is_empty() {
            return;
        }

        let req = wasmrequests::SendRequest {
            address: address.clone(),
            transfer: transfer.to_bytes(),
        };
        sandbox(FN_SEND, &req.to_bytes());
    }

    //fn stateAnchor(&self) -> interface{} {
    //	panic("implement me")
    //}

    // TransferAllowed transfers allowed assets from caller to the specified account
    fn transfer_allowed(&self, agent_id: &ScAgentID, transfer: &ScTransfer) {
        // we need some assets to send
        if transfer.is_empty() {
            return;
        }

        let req = wasmrequests::TransferRequest {
            agent_id: agent_id.clone(),
            transfer: transfer.to_bytes(),
        };
        sandbox(FN_TRANSFER_ALLOWED, &req.to_bytes());
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/sandbox.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {ScAssets, ScBalances, ScTransfer} from "./assets";
import {ScDict, ScImmutableDict} from "./dict";
import {sandbox} from "./host";
import {ScSandboxUtils} from "./sandboxutils";
import {ScImmutableState, ScState} from "./state";
import {ScFunc} from "./contract";
import {CallRequest, DeployRequest, PostRequest, SendRequest, TransferRequest} from "./wasmrequests";
import {requestIDFromBytes, ScRequestID} from "./wasmtypes/screquestid";
import {ScTokenID} from "./wasmtypes/sctokenid";
import {chainIDFromBytes, ScChainID} from "./wasmtypes/scchainid";
import {hashFromBytes, ScHash} from "./wasmtypes/schash";
import {agentIDFromBytes, ScAgentID} from "./wasmtypes/scagentid";
import {ScAddress} from "./wasmtypes/scaddress";
import {Proxy} from "./wasmtypes/proxy";
import {hnameFromBytes, ScHname} from "./wasmtypes/schname";
import {uint64FromBytes} from "./wasmtypes/scuint64";
import {stringToBytes} from "./wasmtypes/scstring";

// @formatter:off
export const FnAccountID              : i32 = -1;
export const FnAllowance              : i32 = -2;
export const FnBalance                : i32 = -3;
export const FnBalances               : i32 = -4;
export const FnBlockContext           : i32 = -5;
export const FnCall                   : i32 = -6;
export const FnCaller                 : i32 = -7;
export const FnChainID                : i32 = -8;
export const FnChainOwnerID           : i32 = -9;
export const FnContract               : i32 = -10;
export const FnDeployContract         : i32 = -11;
export const FnEntropy                : i32 = -12;
export const FnEstimateStorageDeposit : i32 = -13;
export const FnEvent                  : i32 = -14;
export const FnLog                    : i32 = -15;
export const FnMinted                 : i32 = -16;
export const FnPanic                  : i32 = -17;
export const FnParams                 : i32 = -18;
export const FnPost                   : i32 = -19;
export const FnRequest                : i32 = -20;
export const FnRequestID              : i32 = -21;
export const FnRequestSender          : i32 = -22;
export const FnResults                : i32 = -23;
export const FnSend                   : i32 = -24;
export const FnStateAnchor            : i32 = -25;
export const FnTimestamp              : i32 = -26;
export const FnTrace                  : i32 = -27;
export const FnTransferAllowed        : i32 = -28;
export const FnUtilsBech32Decode      : i32 = -29;
export const FnUtilsBech32Encode      : i32 = -30;
export const FnUtilsBlsAddress        : i32 = -31;
export const FnUtilsBlsAggregate      : i32 = -32;
export const FnUtilsBlsValid          : i32 = -33;
export const FnUtilsEd25519Address    : i32 = -34;
export const FnUtilsEd25519Valid      : i32 = -35;
export const FnUtilsHashBlake2b       : i32 = -36;
export const FnUtilsHashName          : i32 = -37;
export const FnUtilsHashSha3          : i32 = -38;
// @formatter:on

// Direct logging of text to host log
export function log(text: string): void {
    sandbox(FnLog, stringToBytes(text));
}

// Direct logging of error to host log, followed by panicking out of the Wasm code
export function panic(text: string): void {
    sandbox(FnPanic, stringToBytes(text));
}

// Direct conditional logging of debug-level informational text to host log
export function trace(text: string): void {
    sandbox(FnTrace, stringToBytes(text));
}

export function paramsProxy(): Proxy {
    return new ScDict(sandbox(FnParams, null)).asProxy();
}

export class ScSandbox {
    // retrieve the agent id of this contract account
    public accountID(): ScAgentID {
        return agentIDFromBytes(sandbox(FnAccountID, null));
    }

    public balance(tokenID: ScTokenID): u64 {
        return uint64FromBytes(sandbox(FnBalance, tokenID.toBytes()));
    }

    // access the current balances for all assets
    public balances(): ScBalances {
        return new ScAssets(sandbox(FnBalances, null)).balances();
    }

    // calls a smart contract function
    protected callWithAllowance(hContract: ScHname, hFunction: ScHname, params: ScDict | null, allowance: ScTransfer | null): ScImmutableDict {
        const req = new CallRequest();
        req.contract = hContract;
        req.function = hFunction;
        if (params === null) {
            params = new ScDict(null);
        }
        req.params = params.toBytes();
        if (allowance === null) {
            allowance = new ScTransfer();
        }
        req.allowance = allowance.toBytes();
        const res = sandbox(FnCall, req.bytes());
        return new ScDict(res).immutable();
    }

    // retrieve the agent id of the owner of the chain this contract lives on
    public chainOwnerID(): ScAgentID {
        return agentIDFromBytes(sandbox(FnChainOwnerID, null));
    }

    // retrieve the hname of this contract
    public contract(): ScHname {
        return hnameFromBytes(sandbox(FnContract, null));
    }

    // retrieve the chain id of the chain this contract lives on
    public currentChainID(): ScChainID {
        return chainIDFromBytes(sandbox(FnChainID, null));
    }

    // logs informational text message
    public log(text: string): void {
        log(text);
    }

    // logs error text message and then panics
    public panic(text: string): void {
        panic(text);
    }

    // retrieve parameters passed to the smart contract function that was called
    public params(): ScImmutableDict {
        return new ScDict(sandbox(FnParams, null)).immutable();
    }

    // panics if condition is not satisfied
    public require(cond: bool, msg: string): void {
        if (!cond) {
            this.panic(msg);
        }
    }

    public results(results: ScDict): void {
        sandbox(FnResults, results.toBytes());
    }

    // deterministic time stamp fixed at the moment of calling the smart contract
    public timestamp(): u64 {
        return uint64FromBytes(sandbox(FnTimestamp, null));
    }

    // logs debugging trace text message
    public trace(text: string): void {
        trace(text);
    }

    // access diverse utility functions
    public utility(): ScSandboxUtils {
        return new ScSandboxUtils();
    }
}

export class ScSandboxView extends ScSandbox {
    // calls a smart contract view
    public call(hContract: ScHname, hFunction: ScHname, params: ScDict | null): ScImmutableDict {
        return this.callWithAllowance(hContract, hFunction, params, null);
    }

    public rawState(): ScImmutableState {
        return new ScImmutableState();
    }
}

export class ScSandboxFunc extends ScSandbox {
    private static entropy: Uint8Array = new Uint8Array(0);
    private static offset: u32 = 0;

    // access the allowance assets
    public allowance(): ScBalances {
        const buf = sandbox(FnAllowance, null);
        return new ScAssets(buf).balances();
    }

    //public blockContext(construct func(sandbox: ScSandbox) interface{}, onClose func(interface{})): interface{} {
    //	panic("implement me")
    //}

    // calls a smart contract function
    public call(hContract: ScHname, hFunction: ScHname, params: ScDict | null, allowance: ScTransfer | null): ScImmutableDict {
        return this.callWithAllowance(hContract, hFunction, params, allowance);
    }

    // retrieve the agent id of the caller of the smart contract
    public caller(): ScAgentID {
        return agentIDFromBytes(sandbox(FnCaller, null));
    }

    // deploys a smart contract
    public deployContract(programHash: ScHash, name: string, description: string, initParams: ScDict | null): void {
        if (initParams === null) {
            initParams = new ScDict(null);
        }
        const req = new DeployRequest();
        req.progHash = programHash;
        req.name = name;
        req.description = description;
        req.params = initParams.toBytes();
        sandbox(FnDeployContract, req.bytes());
    }

    // returns random entropy data for current request.
    public entropy(): ScHash {
        return hashFromBytes(sandbox(FnEntropy, null));
    }

    public estimateStorageDeposit(fn: ScFunc): u64 {
        const req = new PostRequest();
        req.contract = fn.hContract;
        req.function = fn.hFunction;
        req.params = fn.params.toBytes();
        let allowance = fn.allowanceAssets;
        if (allowance === null) {
            allowance = new ScTransfer();
        }
        req.allowance = allowance.toBytes();
        let transfer = fn.transferAssets;
        if (transfer === null) {
            transfer = new ScTransfer();
        }
        req.transfer = transfer.toBytes();
        req.delay = fn.delaySeconds;
        return uint64FromBytes(sandbox(FnEstimateStorageDeposit, req.bytes()));
    }

    // signals an event on the node that external entities can subscribe to
    public event(msg: string): void {
        sandbox(FnEvent, stringToBytes(msg));
    }

    // retrieve the assets that were minted in this transaction
    public minted(): ScBalances {
        return new ScAssets(sandbox(FnMinted, null)).balances();
    }

    // Post (delayed) posts a SC function request
    public post(chainID: ScChainID, hContract: ScHname, hFunction: ScHname, params: ScDict, allowance: ScTransfer, transfer: ScTransfer, delay: u32): void {
        const req = new PostRequest();
        req.chainID = chainID;
        req.contract = hContract;
        req.function = hFunction;
        req.params = params.toBytes();
        req.allowance = allowance.toBytes();
        req.transfer = transfer.toBytes();
        req.delay = delay;
        sandbox(FnPost, req.bytes());
    }

    // generates a random value from 0 to max (exclusive: max) using a deterministic RNG
    public random(max: u64): u64 {
        if (max == 0) {
            this.panic("random: max parameter should be > 0");
        }

        // note that entropy gets reset for every request
        if (ScSandboxFunc.entropy.length == 0) {
            // first time in this: request, initialize with current request entropy
            ScSandboxFunc.entropy = this.entropy().toBytes();
            ScSandboxFunc.offset = 0;
        }
        if (ScSandboxFunc.offset == 32) {
            // ran out of entropy: data, hash entropy for next pseudo-random entropy
            ScSandboxFunc.entropy = this.utility().hashBlake2b(ScSandboxFunc.entropy).toBytes();
            ScSandboxFunc.offset = 0;
        }
        let rnd = uint64FromBytes(ScSandboxFunc.entropy.slice(ScSandboxFunc.offset, ScSandboxFunc.offset + 8)) % max;
        ScSandboxFunc.offset += 8;
        return rnd;
    }

    public rawState(): ScState {
        return new ScState();
    }

    //public request(): ScRequest {
    //	panic("implement me")
    //}

    // retrieve the request id of this transaction
    public requestID(): ScRequestID {
        return requestIDFromBytes(sandbox(FnRequestID, null));
    }

    // retrieve the request sender of this transaction
    public requestSender(): ScAgentID {
        return agentIDFromBytes(sandbox(FnRequestSender, null));
    }

    // Send transfers SC assets to the specified address
    public send(address: ScAddress, transfer: ScTransfer): void {
        // we need some assets to send
        if (transfer.isEmpty()) {
            return;
        }

        const req = new SendRequest();
        req.address = address;
        req.transfer = transfer.toBytes();
        sandbox(FnSend, req.bytes());
    }

    //public stateAnchor(): interface{} {
    //	panic("implement me")
    //}

    // TransferAllowed transfers allowed assets from caller to the specified account
    public transferAllowed(agentID: ScAgentID, transfer: ScTransfer, create: bool): void {
        // we need some assets to send
        if (transfer.isEmpty()) {
            return;
        }

        const req = new TransferRequest();
        req.agentID = agentID;
        req.create = create;
        req.transfer = transfer.toBytes();
        sandbox(FnTransferAllowed, req.bytes());
    }
}
